### ISSUE TITLE
How to reproduce the result of transformer on WMT17 en->de?

### ISSUE THREAD

**Issue by zhao1402072392** [pytorch/fairseq#2652]
Thanks for your contribution. I've tried reproducing the results of the Insertion Transformation on WMT17 En-De translation tasks.

Following the readme from this repository, I've used the training scripts with the following arguments:

cd examples/translation/
bash prepare-iwslt14.sh
cd ../..

TEXT=examples/translation/wmt17_en_de
python3 preprocess.py --source-lang en --target-lang de --dataset-impl raw --trainpref $TEXT/train --validpref $TEXT/valid --testpref $TEXT/test --destdir data-bin/wmt17_en_de

fairseq-train data-bin/wmt17_en_de --dataset-impl raw --arch transformer --optimizer adam --adam-betas '(0.9, 0.98)' --clip-norm 0.0 --lr-scheduler inverse_sqrt --warmup-init-lr 1e-07 --warmup-updates 4000 --lr 0.0007 --min-lr 1e-09 --criterion label_smoothed_cross_entropy --label-smoothing 0.1 --weight-decay 0.0 --max-tokens  6000 --update-freq 2 --no-progress-bar --log-format json --log-interval 50 --save-interval-updates 1000 --keep-interval-updates 20 --fp16 --save-dir checkpoints/wmt17enbpe_debpegitcommmaxtoken6000 

fairseq-generate data-bin/wmt17_en_de --dataset-impl raw --path checkpoints/wmt17enbpe_debpegitcommmaxtoken6000/checkpoint_best.pt --beam 5 --batch-size 200 --lenpen 0.6 --quiet --remove-bpe | tee ./tmp/gen.out

but I got BLEU 22.39 rather than 26.3. Could you give me some suggestions? Thank you.

**Comment by zhao1402072392** [comment]
Could anyone give some suggested commands to train Tansformer in EN-DE's translation task? Thanks

**Comment by risingdhxs** [comment]
Hi,

I haven't tried the "--dataset-impl raw" option so I'm not sure if my experience really applies here, but I would suggest the following:
- add a dropout 0.2 to the training argument: right now it's using the default 0.1 (unless I missed it?)
- weight-decay 0.0001
- max-tokens & update-freq: I'm not sure how many GPUs you are using. I think in another post they mentioned that GPU*max-token*update should be close to 25k. If you're training on 1 gpu, either double the max-tokens or double the update-freq
- inference: in my experience beam=5 and lenpen=0.3 works a bit better
- you also need to do a compound splitting trick after the generation; on WMT En-De it usually gives a boost close to 1 BLEU. See here: https://github.com/pytorch/fairseq/issues/346

Let me know if it helps.

**Comment by zhao1402072392** [comment]
Sorry for taking so long to reply to you, I tried your suggestion and modified the command but it does not work. And then I am trying another command:
nohup python3 train.py data-bin/wmt17_en_de --lr-period-updates 400000 --lr 0.0007 --clip-norm 0.1 --max-tokens 4000 --dataset-impl raw --lr-scheduler triangular --shrink-min --min-lr 1e-20 --max-lr 0.1 --dropout 0.0 --arch transformer --no-progress-bar --weight-decay 0.0001 --criterion label_smoothed_cross_entropy --label-smoothing 0.1  --eval-bleu --eval-bleu-args '{"beam": 5, "max_len_a": 1.2, "max_len_b": 10}'  --eval-bleu-detok moses --eval-bleu-remove-bpe --best-checkpoint-metric bleu --maximize-best-checkpoint-metric --skip-invalid-size-inputs-valid-test --save-dir checkpoints/wmt17enrawdebpemaxtoken4000up40drop00lr7 > wmt17enrawdebpemaxtoken4000up40drop00lr7.out 2>&1 &
But it can only reach 23BLEU

**Comment by RayanMessekher** [comment]
Hi got exactly the same issue as you have. Maybe you've found a solution since then ? What are your results with the same setup on WMT 14 ?
